### PR TITLE
Do not modify saved variables in-place for spectral norm during power iteration

### PR DESCRIPTION
**Pull request by soulitzer**
Interestingly enough, the original code did have a mechanism that aims to prevent this very issue:
but it performs a clone AFTER modifying u and v in-place.
This wouldn't work though because we can later use the cloned u and v in operations that save for backward, and the next time we execute forward, we modify the same cloned u and v in-place.
So if the idea is that we want to avoid modifying saved variable in-place we should clone it BEFORE the in-place operation.

### CODE REVIEW

File: torch/nn/utils/parametrizations.py [hunk 1]
```diff
@@ -84,6 +84,11 @@ def _power_method(self, weight_mat: torch.Tensor, n_power_iterations: int) -> No
 
         # Precondition
         assert weight_mat.ndim > 1
+
+        if n_power_iterations > 0:
+            # See above for why we need this clone
+            self._u = self._u.clone(memory_format=torch.contiguous_format)
```
Comment (albanD) [inline torch/nn/utils/parametrizations.py:90]: I am not sure to understand what's the point of doing the normalize inplace below if we clone here? Why not just make them out of place?

Also the comment above seem to mention that this could be a problem with DP, do we have a test for that? and a test for DP + two forwards?

Comment (soulitzer) [reply torch/nn/utils/parametrizations.py:90]: I don't know what the perf implications are, but if we do multiple iterations of power_iteration, we only do a single out-of-place op if clone.
But since most people have n_power_iterations=1, I can see why it's better to just do out-of-place.
Or I suppose if we really wanted to optimize both cases, we could just add a check: if n_power_iterations > 1, clone. Otherwise, just do it out-of-place.

Yeah from reading the comment in the again, seems like we have to do the modification in-place for DDP to work? I think another alternative is to replace the self.u/self.v with cloned versions AFTER using them in operations that would save them for backward.

It looks like there is a dataparallel test in `test_new_spectral_norm`, we can try to verify this by seeing if CI fails with this current PR.

Comment (Lezcano) [reply torch/nn/utils/parametrizations.py:90]: After a quick search of `torch.nn.utils.spectral_norm(` on github, out of the 30+ occurrences I have seen, none of them uses `n_power_iterations != 1`. I believe that it wouldn't be too bad to remove this optimisation if it's causing so much trouble.

All in all, I cannot think of any application where `n_power_iterations > 2` would be necessary, and making an extra copy in the case  `n == 2` is not too bad.

Comment (albanD) [reply torch/nn/utils/parametrizations.py:90]: I would still like to know if this causes desync in DP or not?

Comment (soulitzer) [reply torch/nn/utils/parametrizations.py]: Looks like this test failure https://app.circleci.com/pipelines/github/pytorch/pytorch/356979/workflows/9e4b4b08-933c-430a-844d-da3c08d10600/jobs/15050920 confirms that we do need to modify u and v in-place for DP to work.

Looks like there is also test that is supposed to test for double forward (and also runs with DP) https://github.com/pytorch/pytorch/blob/9776e1ff2f2658b25c37beaf3f4aea3c8b8bd174/test/test_nn.py#L4210, but it uses .grad underneath (via gradcheck) and computes grad wrt to input only (and not the parameters). The error only reveals itself when we compute grad wrt the actual parameter.

File: torch/nn/utils/parametrizations.py [hunk 2]
```diff
@@ -104,10 +102,17 @@ def forward(self, weight: torch.Tensor) -> torch.Tensor:
             weight_mat = self._reshape_weight_to_matrix(weight)
             if self.training:
                 self._power_method(weight_mat, self.n_power_iterations)
+            # See above on why we need to clone
+            if self.n_power_iterations > 0:
```
Comment (Lezcano) [inline torch/nn/utils/parametrizations.py:106]: `n_power_iterations` is always positive, so this if is unnecessary.
https://github.com/pytorch/pytorch/blob/a3eec40b557db06220af8049fd605e19be39f3ee/torch/nn/utils/parametrizations.py#L22-L24

File: test/test_nn.py [hunk 3]
```diff
@@ -4220,6 +4220,9 @@ def fn(input):
                     out1 = wrapped_m(input)
                     return out0 + out1
 
+                # Make sure we can compute gradients wrt to all the parameters in the case
+                # of double forward
+                fn(input.clone().requires_grad_()).sum().backward()
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_nn.py]: It would be better indeed, but its not possible unless we have a functional version of the nn module

[the remark this answers is not in the document]
Comment (albanD) [reply test/test_nn.py:4225]: Gradcheck is amazing ;) 
You can give it inputs and then use them as captures and it works!
We actually do gradcheck our Modules.
You can use a lambda that takes the parameters as input and just drop them. And give that lambda to gradcheck

[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_nn.py:4225]: Haha I know :P
but I did try something like that...
The issue is I don't think gradcheck modifies the inputs in-place (which wouldn't work anyway for leaf tensors), and instead it produce the perturbed input out-of-place and then call the function again? I recall the numerical Jacobian becoming zero due to this reason
Alternatively I also tried replacing the entire parameter in the lambda every time with the new value, but this also means that 1) we'd need to compute analytical before numerical Jacobian 2) analytical still wouldn't work because the original input we consider the input is no longer part of the graph in the new module forward computation

But maybe I'm missing something - is there a link to where we test our modules with gradcheck?

[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_nn.py:4225]: hmm wait actually it is done in-place

[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_nn.py:4225]: Seems like the reason gradcheck fails is that in training mode (if we do power iteration), power iteration updates _u based on a function of the current weight, which is perturbed by a different one-hot vector every time for finite differencing. This is why having the wrapper function set `_u` to a saved clone does not work. 

Due to this interaction, it is probably not possible to test the double forward in-place issue with gradcheck. (Using eval mode to get gradcheck to work here would skip the in-place update to _u entirely which is what we wanted to check in the first place.)

**Comment by soulitzer** [comment]
As discussed with @albanD offline, since we cannot test with gradcheck with the parametrized spectral norm module as-is, we do some extra testing here. See https://github.com/pytorch/pytorch/pull/63593. The CI should reveal whether it works with DP or not.

This was more complicated that initially thought though because there are many side effects of removing the no_grad. Its very possible we're testing completely different.

Anyway, the things changed in linked PR are:
 - (repeat changes from this PR)
 - remove `@nograd`
 - no longer modify u and v in-place via out arg (because we are in grad mode now), instead we do it out of place then `copy_` after so that DP would still work (though I can't test that locally). We need an extra clone before `F.normalize(...)` because it saves u, v for backward now. Also do the same for `weight_mat`
 - **change clone memory_format=torch.contiguous_format -> perserve_format (TODO: why does this break gradcheck??)**
 - In `test.nn` F.normalize does not support undefined gradients so we need `check_undefined=False` now.
 - Adds `retain_graph` to `.backward` and `.grad` calls because `u` and `v` begin to require grad after the first forward and connects to the graph from the previous iteration

**Comment by albanD** [comment]
> Adds retain_graph to .backward and .grad calls because u and v begin to require grad after the first forward and connects to the graph from the previous iteration

A wrapper function should be used to make sure we don't change the "original" u and v in any way I think.

> In test.nn F.normalize does not support undefined gradients so we need check_undefined=False now.

That is surprising... especially ops with a single output should work out of the box.

**Comment by soulitzer** [comment]
> A wrapper function should be used to make sure we don't change the "original" u and v in any way I think.

This fixed everything. The `check_undefined_grad` issue went away too with the new wrapper. So yeah it's not `F.normalize`, I should've double checked the first time.

**Comment by soulitzer** [comment]
From the linked PR, the test is failing for DP - Not sure if DP is possible when power iteration is done in grad mode because `self._u` needs to be modified in-place, but self._u can be a view apparently. The grad_fn is BroadcastBackward (which I could not grep) seems to return multiple views as well.

**Review by albanD — APPROVED** [review]
ok!
